docs/pipeline_semantics.md

```diff
@@ -0,0 +1,170 @@
+- **Nacked** - a message was negatively acknowledged either by a processor (e.g. a transform failed) or by a
+  destination. If a pipeline contains multiple destinations the message needs to be negatively acknowledged by at least
+  one destination before it is marked as nacked. When a message is nacked the message is passed to
+  the [DLQ](#Dead-letter-queue).
+
+**Important**: if a message gets nacked and the DLQ handler successfully processes the nack (e.g. stores the message in
+the dead letter queue), the source connector will receive an ack as if the message was successfully processed, even
+though Conduit marks it internally as nacked. In other words, the source connector will receive an ack every time
+Conduit handled a message end-to-end and it can be safely discarded from the source.
+
+Pipeline nodes will either leave the message open and send it to the next node for processing or ack/nack it and not
+send it further down the pipeline. If the ack/nack fails, the node will stop running and return an error that will
+consequently stop the whole pipeline. The returned error is stored in the pipeline for further inspection by the user.
+
+### Message state change handlers
+
+A pipeline node can register state change handlers on a message that will be called when the message state changes. This
+is used for example to register handlers that reroute nacked messages to a dead letter queue or to update metrics when a
+message reaches the end of the pipeline.
+
+## Semantics
+
+### Messages are delivered in order
+
+Since messages are passed between nodes in channels and a node only processes one message at a time it is guaranteed
+that messages will flow through the Conduit pipeline in the same order as they are received by the source.
+
+There are two caveats:
+
+- If a pipeline contains multiple source connectors then the order of two messages coming from different connectors is
+  nondeterministic. Messages coming from the same source connector are still guaranteed to retain their order.
+- If a dead letter queue is configured then negatively acknowledged messages will be removed from the stream while the
+  pipeline will be kept running, thus impacting the order of messages.
+
+The order guarantee only holds inside of Conduit. Once a message reaches a destination connector it is allowed to buffer
+messages and batch write them to 3rd party systems. Normally the connector would retain the order, although we can't
+vouch for badly written connectors that don't follow this behavior.
+
+### Messages are delivered at least once
+
+Between pipeline restarts, it is guaranteed that any message that is processed successfully by all nodes and not
+filtered out will be delivered to a destination connector at least once. Multiple deliveries can occur in pipelines with
+multiple destinations that stopped because of a negatively acknowledged record, or pipelines, where a destination
```

Comment:
this time it's an extra comma :D 
```suggestion
multiple destinations that stopped because of a negatively acknowledged record, or pipelines where a destination
```